packages/ts/react-grid/src/autogrid.tsx

```diff
@@ -78,27 +91,108 @@ function createDataProvider<TItem>(
+) {
+  return useCallback((column: any) => {
+    const path = pathFromColumn(column);
+    if (!path) {
+      return null;
+    }
+    const propertyInfo: PropertyInfo = properties.current.find((p) => p.name === path)!;
+
+    return createFilterField(propertyInfo, {
+      onInput: (e: { target: { value: string } }) => {
+        const fieldValue = e.target.value;
+        const filterValue = fieldValue;
+
+        const filter = {
+          propertyId: propertyInfo.name,
+          filterValue,
+          matcher: Matcher.CONTAINS,
+        };
+
+        // eslint-disable-next-line
+        (filter as any).t = 'propertyString';
+        setPropertyFilter.current(filter);
+      },
+    });
+  }, []);
+}
+
+function useHeaderSorterRenderer(properties: React.MutableRefObject<PropertyInfo[]>) {
+  return useCallback((column: any) => {
+    const path = pathFromColumn(column);
+    if (!path) {
+      return null;
+    }
+
+    const propertyInfo: PropertyInfo = properties.current.find((p) => p.name === path)!;
+
+    return <GridSorter path={path}>{propertyInfo.humanReadableName}</GridSorter>;
+  }, []);
+}
+
+function useColumns(
+  model: ModelConstructor<unknown, AbstractModel<unknown>>,
+  setPropertyFilter: React.MutableRefObject<(propertyFilter: PropertyStringFilter) => void>,
+  options: { visibleColumns?: string[]; headerFilters?: boolean },
+) {
   const properties = getProperties(model);
-  const effectiveColumns = visibleColumns ?? properties.map((p) => p.name);
+  const effectiveColumns = options.visibleColumns ?? properties.map((p) => p.name);
   const effectiveProperties = effectiveColumns
     .map((name) => properties.find((prop) => prop.name === name))
     .filter(Boolean) as PropertyInfo[];
+  const propertiesRef = useRef<PropertyInfo[]>([]);
+  propertiesRef.current = properties;
+  const headerFilterRenderer = useHeaderFilterRenderer(propertiesRef, setPropertyFilter);
+  const headerSorterRenderer = useHeaderSorterRenderer(propertiesRef);
 
-  return effectiveProperties.map((p) => (
-    <GridSortColumn path={p.name} header={p.humanReadableName} key={p.name} autoWidth></GridSortColumn>
-  ));
+  return effectiveProperties.map((p) => {
+    if (options.headerFilters) {
+      return (
+        <GridColumnGroup data-path={p.name} key={`group-${p.name}`} headerRenderer={headerSorterRenderer}>
+          <GridColumn path={p.name} headerRenderer={headerFilterRenderer} autoWidth></GridColumn>
+        </GridColumnGroup>
+      );
+    }
+    return <GridColumn path={p.name} headerRenderer={headerSorterRenderer} autoWidth></GridColumn>;
+  });
 }
 
 export function AutoGrid<TItem>({
   service,
   model,
   filter,
   visibleColumns,
+  headerFilters,
   ...gridProps
 }: AutoGridProps<TItem>): JSX.Element {
+  const [internalFilter, setInternalFilter] = useState<AndFilter>({ ...{ t: 'and' }, children: [] });
+
+  const setHeaderPropertyFilter = useRef((propertyFilter: PropertyStringFilter) => {
```

Comment:
This setup has the same issue as the previous one 😅. The ref is only initialized on the first render, and then never updated again. So the function in the ref will always reference the initial `internalFilter` state.

I would suggest to just redefine the function here on every render as before, pass it to `useColumns`, and there store it in a ref on each render.

Reply:
so just like `propertiesRef`...

Reply:
the pattern of
```
  const setPropertyFilterRef = useRef(setPropertyFilter);
  setPropertyFilterRef.current = setPropertyFilter;
```
just looks wrong..

Reply:
Yeah, but that's how it works. The argument to `useRef` will only ever be used in the first call of the hook 🤷‍♂️.